File: 3241-time-taken-to-mark-all-nodes/3241-time-taken-to-mark-all-nodes.cpp
class Solution {
public:
    vector<int> dp;
    vector<vector<int>> adj;
    vector<int> ans;

    void dfs( int node , int par ){
        if( adj[node].size() == 1 && par != -1 ){
            dp[node] = 0;
            return;
        }
        int val = 0;
        for( auto i : adj[node] ){
            if( i == par ) continue;
            dfs( i , node );
            val = max( val , dp[i] + ( i%2 ? 1 : 2 ) );
        }
        dp[node] = val;
    }

    void rerooting( int node, int par ){
        ans[node] = dp[node];

        multiset<int> s;
        for( auto i : adj[node] ){
            s.insert(dp[i] + ( i%2 ? 1 : 2 ));
        }
        int prevnode = dp[node];

        for( auto i : adj[node] ){
            if( i == par ) continue;
            int prev = dp[i];
            int val = dp[i] + ( i%2 ? 1 : 2 );
            s.erase(s.find(val));
            int mx = ( s.size() ? *(--s.end()) : 0 );
            dp[node] = mx;
            dp[i] = max( dp[i], dp[node] + ( node%2 ? 1 : 2 ) );
            rerooting(i,node);
            dp[i] = prev;
            dp[node] = prevnode;
            s.insert(val);
        }
    }


    vector<int> timeTaken(vector<vector<int>>& edges) {
        int n = edges.size() + 1;
        dp.resize(n);
        adj.resize(n);
        ans.resize(n);
        for( auto i : edges ){
            adj[i[0]].push_back(i[1]);
            adj[i[1]].push_back(i[0]);
        }
        dfs(0,-1);
        rerooting(0,-1);
        return ans;
    }
};

File: 3241-time-taken-to-mark-all-nodes/3241-time-taken-to-mark-all-nodes.py
class Solution:
    def timeTaken(self, edges: List[List[int]]) -> List[int]:
        n=len(edges)+1
        m=[[] for _ in range(n)]
        for a,b in edges:
            m[a].append(b)
            m[b].append(a)
        res=[0]*n
        @lru_cache(None)
        def dp(i,prev):
            res=0
            for j in m[i]:
                if j==prev: continue
                res = max(res, dp(j,i))
            res += 2-(i%2)
            # print(i, res)
            return res
        dp(0,0)
        def dp2(i,prev,an):
            x,y = (-1,an), (-2,0)
            for j in m[i]:
                if j==prev: continue
                z = dp(j,i)
                # if i!=0: z+=(2-(prev%2))
                if z>=x[1]: x,y = (j, z), x
                elif z>= y[1]: y = (j,z)
            # print(i, an, x, y)
            res[i] = x[1]
            for j in m[i]:
                if j==prev: continue
                if j!=x[0]: dp2(j, i, x[1] + (2-i%2))
                else: dp2(j, i, y[1] + (2-i%2))
        dp2(0,0,0)
        return res